File: solutions/140_word_break_ii.rb
# Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, add spaces in s to construct a sentence where each word is a valid dictionary word.Return all such possible sentences.
# Note:
#   The same word in the dictionary may be reused multiple times in the segmentation.
#   You may assume the dictionary does not contain duplicate words.
# Example 1:
# Input:
# s = "catsanddog"
# wordDict = ["cat", "cats", "and", "sand", "dog"]
# Output:
# [
#  "cats and dog",
#  "cat sand dog"
# ]
# Example 2:
# Input:
# s = "pineapplepenapple"
# wordDict = ["apple", "pen", "applepen", "pine", "pineapple"]
# Output:
# [
#  "pine apple pen apple",
#  "pineapple pen apple",
#  "pine applepen apple"
# ]
# Explanation: Note that you are allowed to reuse a dictionary word.
# Example 3:
# Input:
# s = "catsandog"
# wordDict = ["cats", "dog", "sand", "and", "cat"]
# Output:
# []
#
#  https://leetcode.com/problems/word-break-ii/description/
require './aux.rb'

# @param {String} s
# @param {String[]} word_dict
# @return {String[]}
def word_break(s, word_dict)
  dict = {}
  backtracking(s, word_dict, dict)
end

def backtracking(s, word_dict, dict)
  return dict[s] if dict.key?(s)
  dict[s] = [] if dict[s].nil?
  word_dict.each do |w|
    dict[s] << s if w == s
    next unless s.start_with?(w)
    backtracking(s[w.length..-1], word_dict, dict).each do |ww|
      dict[s] << w + ' ' + ww
    end
  end
  dict[s]
end

s = 'pineapplepenapple'
word_dict = %w[apple pen applepen pine pineapple]

p word_break(s, word_dict)

File: solutions/140_word_break_ii.cpp
#include "aux.cpp"
/**
Given a non-empty string s and a dictionary wordDict containing a list of
non-empty words, add spaces in s to construct a sentence where each word is a
valid dictionary word.Return all such possible sentences.
Note:
        The same word in the dictionary may be reused multiple times in the
segmentation.
        You may assume the dictionary does not contain duplicate words.
Example 1:
Input:
s = "catsanddog"
wordDict = ["cat", "cats", "and", "sand", "dog"]
Output:
[
 "cats and dog",
 "cat sand dog"
]
Example 2:
Input:
s = "pineapplepenapple"
wordDict = ["apple", "pen", "applepen", "pine", "pineapple"]
Output:
[
 "pine apple pen apple",
 "pineapple pen apple",
 "pine applepen apple"
]
Explanation: Note that you are allowed to reuse a dictionary word.
Example 3:
Input:
s = "catsandog"
wordDict = ["cats", "dog", "sand", "and", "cat"]
Output:
[]

 https://leetcode.com/problems/word-break-ii/description/
 **/
static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();

class Solution {
public:
  vector<string> wordBreak(string s, vector<string> &wordDict) {
    unordered_map<string, vector<string>> dict;
    return backtracking(s, wordDict, dict);
  }

  vector<string> backtracking(string s, vector<string> &wordDict,
                              unordered_map<string, vector<string>> &dict) {
    if (dict.find(s) != dict.end())
      return dict[s];
    for (auto w : wordDict) {
      if (s == w)
        dict[s].push_back(w);
      if (s.substr(0, w.size()) == w) {
        vector<string> tmp = backtracking(s.substr(w.size()), wordDict, dict);
        for (auto ss : tmp)
          dict[s].push_back(w + " " + ss);
      }
    }
    return dict[s];
  }
};

int main() {
  Solution s;
  string ss = "catsanddog";
  vector<string> wordDict = {"cats", "dog", "sand", "and", "cat", "og"};

  ss = "pineapplepenapple";
  wordDict = {"apple", "pen", "applepen", "pine", "pineapple"};

  ss = "aaaaaaa";
  wordDict = {"aaaa", "aa", "a"};
  say(s.wordBreak(ss, wordDict));
}
